File: frontend/user-web/src/app/core/plan.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { RuntimeConfig } from './runtime-config';

export interface Plan {
  planId: string;
  code: string;
  name: string;
  description: string | null;
  price: number;
  durationDays: number;
  storageLimit: number;
  maxUploadSize: number;
  maxVideoDuration: number;
  maxVideoQuality: string | null;
  maxDownloadQuality?: string | null;
  maxMembers: number;
  status: string;
  features?: Record<string, boolean>;
  displayOrder?: number;
}

export interface PlanMember {
  planMemberId: string;
  planId: string;
  ownerUserId: string;
  memberEmail: string;
  memberUserId: string | null;
  status: string;
  invitedAt: string;
  acceptedAt: string | null;
  revokedAt: string | null;
  allocatedStorage?: number | null;
  storageUsed?: number;
}

export interface PlanSubscriptionInfo {
  planHistoryId: string;
  startedAt: string;
  endedAt: string | null;
  autoRenew: boolean;
  isExpired: boolean;
}

export interface MyPlan extends Plan {
  usedStorage: number;
  remainingStorage: number;
  members: PlanMember[];
  subscription?: PlanSubscriptionInfo | null;
  isSharedMember?: boolean;
  activePaidPlanIds?: string[];
  ownerAllocatedStorage?: number | null;
}

export interface PlanShare {
  planId: string;
  code: string;
  name: string;
  description: string | null;
  shareUrl: string;
  status: string;
  features?: Record<string, boolean>;
}

export interface CreatePaymentResponse {
  paymentId: string;
  planId: string;
  planName: string;
  transactionCode: string;
  qrCodeUrl: string;
  amount: number;
  currency: string;
  expiresAt: string;
}

export interface PaymentSummary {
  paymentId: string;
  planId: string;
  planName: string;
  transactionCode: string;
  amount: number;
  currency: string;
  status: string;
  paidAt: string | null;
  createdAt: string;
  expiresAt: string;
}

@Injectable({ providedIn: 'root' })
export class PlanService {
  private http = inject(HttpClient);
  private config = inject(RuntimeConfig);
  private base = this.config.apiBaseUrl + '/plans';
  private paymentsBase = this.config.apiBaseUrl + '/payments';

  getPlans(): Observable<Plan[]> { return this.http.get<Plan[]>(this.base); }
  getPlan(planId: string): Observable<Plan> { return this.http.get<Plan>(`${this.base}/${planId}`); }
  getShare(planId: string): Observable<PlanShare> { return this.http.get<PlanShare>(`${this.base}/${planId}/share`); }
  getMyPlan(): Observable<MyPlan | null> { return this.http.get<MyPlan | null>(`${this.base}/my-plan`); }
  subscribe(planId: string): Observable<Plan> { return this.http.post<Plan>(`${this.base}/${planId}/subscribe`, { autoRenew: false }); }
  invite(email: string, allocatedStorage?: number | null): Observable<PlanMember> { return this.http.post<PlanMember>(`${this.base}/members/invite`, { email, allocatedStorage }); }
  accept(memberId: string, token: string): Observable<PlanMember> { return this.http.post<PlanMember>(`${this.base}/members/${memberId}/accept`, { token }); }
  revoke(memberId: string): Observable<void> { return this.http.delete<void>(`${this.base}/members/${memberId}`); }
  updateMemberStorage(memberId: string, allocatedStorage: number | null): Observable<PlanMember> {
    return this.http.patch<PlanMember>(`${this.base}/my-subscription/members/${memberId}/storage`, { allocatedStorage });
  }
  updateOwnerStorage(allocatedStorage: number | null): Observable<void> {
    return this.http.patch<void>(`${this.base}/my-subscription/owner-storage`, { allocatedStorage });
  }

  initiatePayment(planId: string, autoRenew = false, idempotencyKey?: string): Observable<CreatePaymentResponse> {
    return this.http.post<CreatePaymentResponse>(this.paymentsBase, { planId, autoRenew, idempotencyKey });
  }

  getPayment(paymentId: string): Observable<PaymentSummary> {
    return this.http.get<PaymentSummary>(`${this.paymentsBase}/${paymentId}`);
  }

  getMyPayments(): Observable<PaymentSummary[]> {
    return this.http.get<PaymentSummary[]>(this.paymentsBase);
  }
}

File: mobile/user-app/lib/features/plans/plan_service.dart
import '../../auth.dart';

class PlanService {
  PlanService(this.auth);
  final AuthController auth;

  Future<List<Map<String, dynamic>>> catalogue() async =>
      (await auth.api.requestList('GET', '/plans'))
          .whereType<Map>()
          .map((item) => Map<String, dynamic>.from(item))
          .toList();

  Future<Map<String, dynamic>?> myPlan() async {
    final value = await auth.protected('GET', '/plans/my-plan');
    return value.isEmpty ? null : value;
  }

  Future<Map<String, dynamic>> share(String planId) =>
      auth.api.request('GET', '/plans/${Uri.encodeComponent(planId)}/share');

  Future<Map<String, dynamic>> get(String planId) =>
      auth.api.request('GET', '/plans/${Uri.encodeComponent(planId)}');

  Future<Map<String, dynamic>> subscribe(
    String planId, {
    bool autoRenew = false,
  }) => auth.protected(
    'POST',
    '/plans/${Uri.encodeComponent(planId)}/subscribe',
    body: {'autoRenew': autoRenew},
  );

  Future<Map<String, dynamic>> invite(String email, {int? allocatedStorage}) =>
      auth.protected(
        'POST',
        '/plans/members/invite',
        body: {
          'email': email.trim(),
          if (allocatedStorage != null) 'allocatedStorage': allocatedStorage,
        },
      );

  Future<Map<String, dynamic>> acceptInvitation(
    String memberId,
    String token,
  ) => auth.protected(
    'POST',
    '/plans/members/${Uri.encodeComponent(memberId)}/accept',
    body: {'token': token},
  );

  Future<void> removeMember(String memberId) async {
    await auth.protected(
      'DELETE',
      '/plans/members/${Uri.encodeComponent(memberId)}',
    );
  }

  Future<Map<String, dynamic>> updateMemberStorage(
    String memberId,
    int? allocatedStorage,
  ) => auth.protected(
    'PATCH',
    '/plans/members/${Uri.encodeComponent(memberId)}/storage',
    body: {'allocatedStorage': allocatedStorage},
  );

  Future<void> updateOwnerStorage(int? allocatedStorage) async {
    await auth.protected(
      'PATCH',
      '/plans/my-subscription/owner-storage',
      body: {'allocatedStorage': allocatedStorage},
    );
  }
}
